Job_SeparateClusterTest.stressTest_getJobStatus_client failure

_master_ (commit bc8018a5c173f3df04e94ee9de5d7b5ebb5ad52b)

Failed on PR builder: http://jenkins.hazelcast.com/job/jet-oss-pr-builder/3180/testReport/junit/com.hazelcast.jet.core/Job_SeparateClusterTest/stressTest_getJobStatus_client/

Stacktrace:
```
java.util.concurrent.ExecutionException: com.hazelcast.jet.core.JobNotFoundException: Job with id 04e7-aa14-6940-0001 not found
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at com.hazelcast.jet.core.Job_SeparateClusterTest.stressTest_getJobStatus(Job_SeparateClusterTest.java:211)
	at com.hazelcast.jet.core.Job_SeparateClusterTest.stressTest_getJobStatus_client(Job_SeparateClusterTest.java:164)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)
	at com.hazelcast.test.DumpBuildInfoOnFailureRule$1.evaluate(DumpBuildInfoOnFailureRule.java:37)
	at com.hazelcast.test.jitter.JitterRule$1.evaluate(JitterRule.java:104)
	at com.hazelcast.test.metrics.MetricsRule$1.evaluate(MetricsRule.java:62)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.jet.core.JobNotFoundException: Job with id 04e7-aa14-6940-0001 not found
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at com.hazelcast.internal.util.ExceptionUtil.tryWrapInSameClass(ExceptionUtil.java:233)
	at com.hazelcast.internal.util.ExceptionUtil.wrapException(ExceptionUtil.java:138)
	at com.hazelcast.internal.util.ExceptionUtil.peel(ExceptionUtil.java:114)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.peelIfNeeded(AbstractMessageTask.java:348)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.encodeException(AbstractMessageTask.java:275)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.sendClientMessage(AbstractMessageTask.java:269)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.handleProcessingFailure(AbstractMessageTask.java:194)
	at com.hazelcast.client.impl.protocol.task.AbstractAsyncMessageTask.sendResponseOrHandleFailure(AbstractAsyncMessageTask.java:84)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456)
	at com.hazelcast.internal.util.ConcurrencyUtil$1.execute(ConcurrencyUtil.java:39)
	at java.util.concurrent.CompletableFuture$UniCompletion.claim(CompletableFuture.java:543)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:765)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.cascadeException(AbstractInvocationFuture.java:1296)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.access$200(AbstractInvocationFuture.java:64)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$ApplyNode.execute(AbstractInvocationFuture.java:1438)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.unblockOtherNode(AbstractInvocationFuture.java:739)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.unblockAll(AbstractInvocationFuture.java:700)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.unblockAll(AbstractInvocationFuture.java:697)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.complete0(AbstractInvocationFuture.java:1191)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.completeExceptionallyInternal(AbstractInvocationFuture.java:1179)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.completeExceptionally(Invocation.java:682)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.notifyThrowable(Invocation.java:392)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.notifyError(Invocation.java:336)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.sendResponse(Invocation.java:233)
	at com.hazelcast.spi.impl.operationservice.Operation.sendResponse(Operation.java:470)
	at com.hazelcast.jet.impl.operation.AsyncOperation.doSendResponse(AsyncOperation.java:80)
	at com.hazelcast.jet.impl.operation.AsyncOperation.lambda$run$0(AsyncOperation.java:59)
	at com.hazelcast.jet.impl.util.ExceptionUtil.lambda$withTryCatch$0(ExceptionUtil.java:182)
	at com.hazelcast.spi.impl.DeserializingCompletableFuture$DeserializingBiConsumer.accept(DeserializingCompletableFuture.java:394)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:457)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: com.hazelcast.jet.core.JobNotFoundException: Job with id 04e7-aa14-6940-0001 not found
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$callWithJob$34(JobCoordinationService.java:633)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$43(JobCoordinationService.java:999)
	at com.hazelcast.internal.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:64)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)
```

Standard output:
```
java.lang.NullPointerException: null
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$toNullFunction$33(JobCoordinationService.java:575) ~[classes/:?]
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$callWithJob$34(JobCoordinationService.java:623) ~[classes/:?]
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$43(JobCoordinationService.java:999) ~[classes/:?]
	at com.hazelcast.internal.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:64) [hazelcast-4.1-SNAPSHOT.jar:4.1-SNAPSHOT]
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217) [hazelcast-4.1-SNAPSHOT.jar:4.1-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_251]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_251]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_251]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76) [hazelcast-4.1-SNAPSHOT.jar:4.1-SNAPSHOT]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102) [hazelcast-4.1-SNAPSHOT.jar:4.1-SNAPSHOT]
```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java
@@ -360,7 +360,7 @@ public class JobCoordinationService {
                     }
                     logger.fine("Ignoring cancellation of a completed job " + idToString(jobId));
                 },
-                jobExecutionRecord -> {
+                jobRecord -> {
                     // we'll eventually learn of the job through scanning of records or from a join operation
                     throw new RetryableHazelcastException("No MasterContext found for job " + idToString(jobId) + " for "
                             + terminationMode);
@@ -416,7 +416,7 @@ public class JobCoordinationService {
                             : jobStatus;
                 },
                 JobResult::getJobStatus,
-                null,
+                jobRecord -> NOT_RUNNING,
                 jobExecutionRecord -> jobExecutionRecord.isSuspended() ? SUSPENDED : NOT_RUNNING
         );
     }
@@ -433,7 +433,7 @@ public class JobCoordinationService {
                     List<RawJobMetrics> metrics = jobRepository.getJobMetrics(jobId);
                     cf.complete(metrics != null ? metrics : emptyList());
                 },
-                record -> cf.complete(emptyList())
+                jobRecord -> cf.complete(emptyList())
         );
         return cf;
     }
@@ -457,7 +457,7 @@ public class JobCoordinationService {
                 jobResult -> {
                     throw new IllegalStateException("Job already completed");
                 },
-                jobExecutionRecord -> {
+                jobRecord -> {
                     throw new RetryableHazelcastException("Job " + idToString(jobId) + " not yet discovered");
                 }
         );
@@ -560,17 +560,22 @@ public class JobCoordinationService {
             long jobId,
             @Nonnull Consumer<MasterContext> masterContextHandler,
             @Nonnull Consumer<JobResult> jobResultHandler,
-            @Nullable Consumer<JobExecutionRecord> jobExecutionRecordHandler
+            @Nonnull Consumer<JobRecord> jobRecordHandler
     ) {
         return callWithJob(jobId,
                 toNullFunction(masterContextHandler),
                 toNullFunction(jobResultHandler),
-                toNullFunction(null),
-                toNullFunction(jobExecutionRecordHandler)
+                toNullFunction(jobRecordHandler),
+                null
         );
     }
 
-    private <T, R> Function<T, R> toNullFunction(Consumer<T> consumer) {
+    /**
+     * Returns a function that passes its argument to the given {@code
+     * consumer} and returns {@code null}.
+     */
+    @Nonnull
+    private <T, R> Function<T, R> toNullFunction(@Nonnull Consumer<T> consumer) {
         return val -> {
             consumer.accept(val);
             return null;
@@ -581,13 +586,10 @@ public class JobCoordinationService {
             long jobId,
             @Nonnull Function<MasterContext, T> masterContextHandler,
             @Nonnull Function<JobResult, T> jobResultHandler,
-            @Nullable Function<JobRecord, T> jobRecordHandler,
+            @Nonnull Function<JobRecord, T> jobRecordHandler,
             @Nullable Function<JobExecutionRecord, T> jobExecutionRecordHandler
     ) {
         assertIsMaster("Cannot do this task on non-master. jobId=" + idToString(jobId));
-        if (jobRecordHandler == null && jobExecutionRecordHandler == null) {
-            throw new IllegalArgumentException();
-        }
 
         return submitToCoordinatorThread(() -> {
             // when job is finalized, actions happen in this order:
@@ -619,7 +621,7 @@ public class JobCoordinationService {
                 return jobExecutionRecordHandler.apply(jobExRecord);
             }
             JobRecord jobRecord;
-            if (jobRecordHandler != null && (jobRecord = jobRepository.getJobRecord(jobId)) != null) {
+            if ((jobRecord = jobRepository.getJobRecord(jobId)) != null) {
                 return jobRecordHandler.apply(jobRecord);
             }
 

